core/src/main/scala-2/caliban/schema/ArgBuilderDerivation.scala

```diff
@@ -54,6 +57,24 @@ trait CommonArgBuilderDerivation {
         }
       case None        => Left(ExecutionError(s"Can't build a trait from input $input"))
     }
+
+  private def makeOneOffBuilder[A](ctx: SealedTrait[ArgBuilder, A]): ArgBuilder[A] =
+    new ArgBuilder[A] {
+      private lazy val builders = ctx.subtypes.map(_.typeclass)
+
+      def build(input: InputValue): Either[ExecutionError, A] = input match {
+        case InputValue.ObjectValue(fields) if fields.size == 1 =>
+          builders.view
+            .map(_.build(input))
+            .find(_.isRight)
+            .getOrElse(Left(ExecutionError(s"Invalid oneOf input $fields for trait ${ctx.typeName.short}}")))
```

Comment:
Open to suggestions on how we can make this more efficient. I can't figure out a way to extract the field argument here unfortunately :(

Reply:
Isn't there a possibility that some types can successfully be parsed as another type of the oneof?

Reply:
We get the subclass name though right? I think we could recapitalize the type name `Dog => dog` as the default strategy, and allow people to override that explicitly with an annotation

Reply:
> Isn't there a possibility that some types can successfully be parsed as another type of the oneof?

@ghostdogpr No because we ensure that the argument names of the leaf cases are unique

>  I think we could recapitalize the type name Dog => dog as the default strategy, and allow people to override that explicitly with an annotation

@paulpdaniels  I think this could work, but just to be on the same page. Let's say we want to implement this:

```gql
input BarInput {
  a: String!
  b: Int!
}

input FooInput @oneOf {
  stringArg: String
  barArg: BarInput
}
```

Currently, we can do it as:

```scala
case class Bar(a: String, b: Int)

sealed trait Foo
object Foo:
  case class ArgA(stringArg: String) extends Foo
  case class ArgB(barArg: Bar) extends Foo
```
With your recommendation @paulpdaniels, we need to do the following:

```scala
sealed trait Foo
object Foo:
  @GQLValueType
  @GQLName("stringArg")  // Otherwise it'll be named ArgA
  case class ArgA(stringArg: String) extends Foo
  
  @GQLName("barArg")
  case class Bar(a: String, b: Int) extends Foo

  // Alternatively:
  @GQLInputName("BarInput")
  case class BarArg(a: String, b: Int) extends Foo
```

If this is what you meant, I like it. It reuses the existing annotations, and prevents having to create an extra case class as in the case of `case class ArgB(barArg: Bar) extends Foo` in my first example

Reply:
`GQLName`is already used for renaming fields though

Reply:
@ghostdogpr are you suggesting we should use `GQLName` to rename the fields here as well? My only concern is that in this case `GQLName` annotates the case class, so if someone wants to use this `sealed trait` as a union, then the type will not be named properly

Reply:
Actually I think I prefer the initial way:
```scala
case class Bar(a: String, b: Int)

sealed trait Foo
object Foo {
  case class ArgA(stringArg: String) extends Foo
  case class ArgB(barArg: Bar) extends Foo
}
```

That's how arguments work currently and I think it's good to be consistent.

Reply:
From a UX perspective, I prefer this as well. However, this approach makes 2 things a bit more complicated:

1. We have to iterate through the leaf `ArgBuilder`s in order to find which one matches the input as per my comment above. Personally, I don't think this is much of an issue. We could even do something clever and memoize this after the iteration succeeds for the first time.
2. We have to ensure that `ArgA`, `ArbB` etc have only 1 argument. Unfortunately, the only time that we can do that is during derivation as we need to flatten the list of arguments when we create the schema. This probably means we need to decide on what's best UX wise for this case. We can either:
    1. Throw an error during derivation (perhaps use macros in Scala 3?). This is not very pure but it's safe as the error will be thrown during the application startup
    2. Drop any extra arguments from the schema, but since we can't do any validation this will likely result in unexpected runtime errors
    3. ???

Reply:
Annoyingly, this would be the easiest way to define a schema / arg builder for a `@oneOf` input, but then it's really annoying to extract the argument that was passed:

```scala
@GQLOneOfInput
case class Foo(
  stringArg: Option[String],
  barArg: Option[Bar]
)
```

Reply:
Couldn't we use `__Type#origin` to find during validation that 2 fields come from the same object and fail in that case?

Reply:
Unfortunately no. For Scalars, we don't set a value for the origin field. I actually tried setting it manually only for oneOf input arguments and it didn't work either, it caused issues with the `validateClashingTypes` validation.

The only way I could figure out how to propagate the type info was to add a field on `__InputValue` and exclude that field from any schema generation: https://github.com/ghostdogpr/caliban/pull/1846/files#diff-7fd5a1e4c9e85d511a507e4a9d0ddb63b131a5be6decb8feb3f95d0c00e5f31cR16

It's a bit of a hack unfortunately, but it works. Other than binary compatibility (which we're breaking with this PR anyways), do we see any other issues with doing this?

Reply:
Seems okay to me, I don't mind that hack so much if things stay simple on the user side.

Reply:
I also rewrote the `ArgBuilder`s for OneOf inputs to fallbacks `orElse`. Under the hood it's still the same as iterating through them I suppose but at least this feels more Caliban-y 😅